Screen handler API

Juuxel opened the pull request:
<img src="https://cdn.discordapp.com/attachments/521545796882006027/700409365558984704/unknown.png" alt="a real screen handler" height=256> Closes #175.

Adds a screen handler API based on `ScreenHandlerType`s.

- The `ScreenHandlerType` constructor is private in vanilla. This API adds two new methods for creating the types:
  - `ScreenHandlerRegistry.registerSimple()` creates and registers the type for a simple vanilla-like screen handler.
  - `ScreenHandlerRegistry.registerExtended()` creates and registers the type for an *extended* screen handler. Extended screen handlers can send additional data to the client when they are opened.
- `ScreenRegistry` replaces the registration methods of vanilla's `HandledScreens`, which are private.
- Because these screen handlers are type-based, they can be opened with vanilla's method: `ServerPlayerEntity.openHandledScreen(NamedScreenHandlerFactory)`.
  - Extended screen handlers have their own factory interface, `ExtendedScreenHandlerFactory`, that extends `NamedScreenHandlerFactory`. It defines one extra method for writing the additional data sent to the client into a packet buffer.
- The old fabric-containers-v0 module is deprecated. Due to it not tracking the types of screen handlers, it can't be ported cleanly on top of this module.

This PR is mirrored in [the FabLabs Screen Handlers project](https://github.com/FabLabsMC/ScreenHandlers), and it can be downloaded as a standalone library from there. The FabLabs repository also includes [a test/example mod](https://github.com/FabLabsMC/ScreenHandlers/tree/1.16/example/src/main/java/io/github/fablabsmc/fablabs/impl/screenhandler/example).

Juuxel committed:
Add screen handler API

Juuxel committed:
Use more descriptive letters for FabricHandledScreens generics

Juuxel committed:
Add missing Deprecated annotations

Juuxel committed:
Bump fabric-containers-v0 version

Juuxel committed:
Remove unused ServerPlayerEntityAccessor

i509VCB commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/impl/screenhandler/ExtendedScreenHandlerType.java:
```diff
@@ -0,0 +1,46 @@
+
+	@Environment(EnvType.CLIENT)
+	@Override
+	public T create(int syncId, PlayerInventory inventory) {
```
I like final things (could you make this method final?)

Juuxel replied:
It doesn't really matter as the class is final.

liach commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/api/screenhandler/v1/ScreenHandlers.java:
```diff
@@ -0,0 +1,91 @@
+/**
+ * An utility for creating screen handler types.
+ */
+public interface ScreenHandlers {
```
This sounds too much like a utility class... You can just make this a utility class which methods accomplish the functionalities (allows preserving binary compatibility across different implementations/updates as well)

liach commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/api/screenhandler/v1/client/FabricHandledScreens.java:
```diff
@@ -0,0 +1,65 @@
+ * An utility for registering screens with screen handlers.
+ */
+@Environment(EnvType.CLIENT)
+public interface FabricHandledScreens {
```
Class sounds too much like a utility class

i509VCB commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/impl/screenhandler/client/NetworkingClient.java:
```diff
@@ -0,0 +1,85 @@
+		ScreenHandlerType<?> type = Registry.SCREEN_HANDLER.get(typeId);
+
+		if (type == null) {
+			LOGGER.warn("[Fabric] Unknown screen handler ID: {}", typeId);
```
Unsure about putting the `[Fabric]` in the logs directly. I want to get someone else to look at this as a standards question as well.

liach replied:
Good spotting. The logger name should suffice; this `[Fabric]` should be removed.

i509VCB reviewed (commented):
Everything impl side looks fine to me

liach commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/impl/screenhandler/client/NetworkingClient.java:
```diff
@@ -0,0 +1,85 @@
+		int typeId = buf.readVarInt();
+		int syncId = buf.readVarInt();
+		Text title = buf.readText();
+		ScreenHandlerType<?> type = Registry.SCREEN_HANDLER.get(typeId);
```
Hmm, is `Registry.get` safe in netty eventloop

Juuxel replied:
I think it's also done in vanilla?

liach replied:
This one should be fine assuming registry isn't changed when the server/client play is active, not so much for client player though if there is a disconnection etc.

Juuxel replied:
The type also needs to be there for the deserialization of the extra data in `buf`.

liach replied:
I mean the access to client player below is unsafe.

Juuxel replied:
![](https://user-images.githubusercontent.com/6596629/79682649-90ef5f80-822c-11ea-81be-8b400eb66b53.png)
Would this be a solution for the thread issue? (and do I have to manually `release` the slice or does the GC do its job)

liach replied:
You create the retained slice outsdie of the runnable. Otherwise when the handler's method exits the buf is released and by the time you call `retainedSlice` on buf on main thread, the buf is gone

liach replied:
In fact, it seems you can just call `buf.retain` before you call `execute` and just pass the `buf` to `openScreen`

liach commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/impl/screenhandler/client/NetworkingClient.java:
```diff
@@ -0,0 +1,85 @@
+			MinecraftClient client = MinecraftClient.getInstance();
+			PlayerEntity player = client.player;
+
+			Screen screen = screenFactory.create(
```
similar concerns here, some read accesses may be unsafe in event loop

Juuxel committed:
Remove [Fabric] from log messages

shedaniel commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/mixin/screenhandler/ServerPlayerEntityMixin.java:
```diff
@@ -0,0 +1,80 @@
+	private final ThreadLocal<ScreenHandler> fabric_openedScreenHandler = new ThreadLocal<>();
+
+	@Inject(method = "openHandledScreen(Lnet/minecraft/screen/NamedScreenHandlerFactory;)Ljava/util/OptionalInt;", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/network/ServerPlayNetworkHandler;sendPacket(Lnet/minecraft/network/Packet;)V"), locals = LocalCapture.CAPTURE_FAILHARD)
+	private void fabric_onOpenHandledScreen_cacheScreenHandler(NamedScreenHandlerFactory factory, CallbackInfoReturnable<OptionalInt> info, ScreenHandler handler) {
```
Are we going for prefixing the inject methods with `fabric` and `_`? I am seeing mostly no prefix _on inject_, this is an inject so conflicting names shouldn't matter right?

Juuxel replied:
The fabric_ prefix is also for exception stacktraces because `VanillaClass.someMethod$onSomeMethod` doesn't tell what mod caused the error, unlike `fabric_onSomeMethod`.

shedaniel replied:
We might want to have a standard in place for this, 99% of mixins have no prefix right now, we might want to refactor all of them if that's what we are going for.

shedaniel replied:
But that's a topic for another day, let's not bloat this thing

i509VCB replied:
That is actually an issue I brought up already https://github.com/FabricMC/fabric/issues/510

LemmaEOF replied:
Prefixes are not necessary on `@Inject` methods, since Mixin prepends special signature info onto each method. For `MixinAbstractFurnaceBlockEntity`'s [`fuelTimeMapHook`](https://github.com/FabricMC/fabric/blob/1.16/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/mixin/content/registry/MixinAbstractFurnaceBlockEntity.java#L34) method, Mixin names it `handler$zeh000$fuelTimeMapHook`. It's also given a `@MixinMerged` annotation noting the mixin class that added the injection. If the `fabric_` prefix here is because Mixin error handling doesn't expose the cause well enough, then we can just improve error handling on our Mixin fork. This method should just be named `cacheScreenHandler`.

liach replied:
we should use prefix, though i personally prefer `fabric$` prefix as `_` looks too ugly

LemmaEOF replied:
We should follow the standard in the rest of Fabric API or create a new PR to enforce the `fabric$` prefix standard across the rest of the API. Renaming mixin methods is not breaking, so it can be done safely at any time. Stick with the API convention for now.

Juuxel replied:
There's no API convention, though. Fabric uses all of those different styles that i5 listed in #510.

LemmaEOF replied:
oh god, really? I could have sworn there was one already

shedaniel replied:
I searched through some mixins and 99% of them have no prefix, the example in i5's issue for `fabric_` is not an inject, most other `fabric_` are overrides, the current majority is no prefix.

I am not saying that no prefix is the best style but mostly injects in fabric API have no prefixes right now, and we should stick to the masses until we have a PR to change this.

shedaniel commented on fabric-screen-handler-api-v1/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,30 @@
+  ],
+  "depends": {
+    "fabricloader": ">=0.8.0",
+    "fabric-api-base": "*"
```
Since I see the `fabric-networking-v0` in the build.gradle, care putting it here as well

Juuxel replied:
Done.

Juuxel committed:
Convert FabricHandledScreens to a class

Juuxel committed:
Add dependency on networking-v0

Juuxel committed:
Resolve threading issues with NetworkingClient

Juuxel committed:
Use Identifiers instead of int IDs, make ScreenHandlers a class

Also added a private constructor to FabricHandledScreens.

Juuxel committed:
NetworkingClient -> ClientNetworking

Juuxel committed:
Use more descriptive names for injections

Juuxel commented:
@liach I made those interfaces classes instead (and moved the packet sending method to `Networking`).

Juuxel committed:
Use ID parameter instead of fetching again

Juuxel committed:
Add clarifying comment

modmuss50 commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/mixin/screenhandler/ServerPlayerEntityMixin.java:
```diff
@@ -0,0 +1,81 @@
+	private int screenHandlerSyncId;
+
+	@Unique
+	private final ThreadLocal<ScreenHandler> fabric_openedScreenHandler = new ThreadLocal<>();
```
Whats the use of a ThreadLocal?

modmuss50 replied:
Edit: Oh I see.

modmuss50 commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/mixin/screenhandler/ServerPlayerEntityMixin.java:
```diff
@@ -0,0 +1,81 @@
+	@Redirect(method = "openHandledScreen(Lnet/minecraft/screen/NamedScreenHandlerFactory;)Ljava/util/OptionalInt;", at = @At(value = "INVOKE", target = "Lnet/minecraft/server/network/ServerPlayNetworkHandler;sendPacket(Lnet/minecraft/network/Packet;)V"))
+	private void fabric_replaceVanillaScreenPacket(ServerPlayNetworkHandler networkHandler, Packet<?> packet, NamedScreenHandlerFactory factory) {
+		if (factory instanceof ExtendedScreenHandlerFactory) {
+			ScreenHandler handler = fabric_openedScreenHandler.get();
```
Would it be a good idea to set `fabric_openedScreenHandler` value to null here? Save it hanging around?

Juuxel replied:
It's cleared at `RETURN`, but I think I can move it there.

Player3324 commented:
What is the rationale here? How does it improve upon the old API?

As far as I can tell it mostly brings everything closer to vanilla. This may actually be worse through increased exposure to their changes and potentially missing more mod-centric features. For example passing PlayerInventory instead of PlayerEntity hides the player instance that may be used for more complex GUIs that use external player specific API. Passing the title over the network is another peculiarity that moves part of the presentation to the server without an obvious strong reason.

Earthcomputer commented:
Passing the title over the network allows for the container block's/entity's custom name to display in the GUI

Juuxel commented:
> What is the rationale here?

Firstly, to improve the naming. `fabric-containers-v0` uses the old and bad name `Container`, while this module uses the current (better) Yarn name.

> How does it improve upon the old API?

By using a system closer to vanilla, the user-side code becomes a lot cleaner, especially around opening menus. Instead of using a custom `ContainerProviderRegistry.open` method, you can just use vanilla's method on players: (block example below)
```java
player.openHandledScreen(state.createScreenHandlerFactory(world, pos));
```

Users also don't have to mess with a packet buffer that isn't needed for most types of screens. Chest-like containers and furnace-like machines (like the basic machines in TR) don't need any custom data to be sent over.

> As far as I can tell it mostly brings everything closer to vanilla. This may actually be worse through increased exposure to their changes and potentially missing more mod-centric features.

This API has all features that the old container module has. Mojang also doesn't seem to be refactoring screens and menus much, considering their code has mostly stayed the same across versions.

> For example passing PlayerInventory instead of PlayerEntity hides the player instance that may be used for more complex GUIs that use external player specific API.

This is a small detail that can be changed (doesn't really matter to me tbh), and you can already convert between the two as they contain references to each other.

> Passing the title over the network is another peculiarity that moves part of the presentation to the server without an obvious strong reason.

As @Earthcomputer said, it's meant for the main use of menus: container blocks. All containers in vanilla can be renamed using anvils, which makes the title part of the vanilla packet protocol.

Player3324 commented:
Thanks for the info, this works for me.

Regarding the PR itself: I don't like having a public class with a generic name such as Factory in the API, even if it is nested, since this is asking for trouble with automatic import handling. The distinction between ScreenHandlers, FabricHandledScreens and their contents is very confusing, their Javadoc description is not too helpful. If they are just for convenience, it'd be nice to state so explicitly instead of relying on a certain interpretation of "utility". It is important to tell the "why" just as much as the "what", think "ok so this allows me to create a screen handler type, but why would i do that? what do i have to do at all to get anyhwhere? which of this is mandatory and what are the alternatives?"

Class Javadoc should add example code and refer to it from other related classes, this kind of example targets advanced devs that just aren't familiar with the specific API (Fabric or MC), but need a few pointers to get going properly+quickly. The wiki goes beyond that, but shouldn't be mandatory.

The package is incorrect and checkstyle should reject it, `net.fabricmc.fabric.api.screenhandler.v1.client` needs to be `net.fabricmc.fabric.api.client.screenhandler.v1`. Please test that it is actually being rejected and adjust.

How useful is `ExtendedScreenHandlerFactory.of`? When/why would it be used?

Doesn't the lack of context in `ExtendedScreenHandlerFactory.writeScreenData` cause serious issues? I can trivially imagine some cases where those contents might be specific to the player.

immibis commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/api/screenhandler/v1/ExtendedScreenHandlerFactory.java:
```diff
@@ -0,0 +1,68 @@
+	/**
+	 * Writes additional server -> client screen opening data to the buffer.
+	 */
+	void writeScreenData(PacketByteBuf buf);
```
Since the comment says "screen opening data", maybe rename to writeScreenOpeningData

immibis commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/api/screenhandler/v1/ExtendedScreenHandlerFactory.java:
```diff
@@ -0,0 +1,68 @@
+	 * @param packetWriter the packet writer
+	 * @return the created factory
+	 */
+	static ExtendedScreenHandlerFactory of(NamedScreenHandlerFactory factory, Consumer<PacketByteBuf> packetWriter) {
```
What is the use case for this function?

Juuxel replied:
If you already have a `NamedScreenHandlerFactory` from vanilla, you can add the extra screen opening data with a lambda. Tbh, I didn't find any use cases for this in the example mod, so I might just remove it.

immibis replied:
Can you think of any case where you might have a NamedScreenHandlerFactory from vanilla but also screen opening data? It seems like you'd write an ExtendedScreenHandlerFactory in that case because the vanilla NSHF is going to be completely useless for your custom screen, because it's going to open a vanilla screen.

liach replied:
Mostly when a vanilla named screen handler is used the data needed by the screen is already on client's block entity or so

Juuxel replied:
I already removed it in the FabLabs repo. I'll just have to port all the changes here.

Juuxel commented:
@sfPlayer1 

> The package is incorrect and checkstyle should reject it, net.fabricmc.fabric.api.screenhandler.v1.client needs to be net.fabricmc.fabric.api.client.screenhandler.v1. Please test that it is actually being rejected and adjust.

The PR passes the checkstyle so I didn't notice that before. Will fix.

> Doesn't the lack of context in ExtendedScreenHandlerFactory.writeScreenData cause serious issues? I can trivially imagine some cases where those contents might be specific to the player.

Good point, I'll add a player parameter (and you can get the world from there if you don't have it already).

> I don't like having a public class with a generic name such as Factory in the API,

Hm, maybe `ScreenHandlers.SimpleClientHandlerFactory` and `.ExtendedClientHandlerFactory`? I'm open to suggestions on these.

I'll also be improving the docs, thanks for the feedback on those.

Juuxel committed:
Port changes from FabLabs

Juuxel committed:
Fix compilation

Juuxel commented:
Okay, I ported all of the changes I made to the FabLabs project today.

Summary:
- `ScreenHandlers` -> `FabricScreenHandlerTypes`
  - Works exclusively with types.
  - According to Player, the -s suffix is not enough to distinguish it from vanilla classes.
  - The creation methods now take an `Identifier` and register the type as well.
  - `SimpleFactory`, `ExtendedFactory` -> `SimpleClientHandlerFactory`, `ExtendedClientHandlerFactory`
    - Clarifies the meaning especially when working with IDEs' autocomplete.
- Moved `FabricHandledScreens` to `api.client.screenhandler.v1`
- `ExtendedScreenHandlerFactory`
  - Removed `of` as it didn't have a proper use.
  - `writeScreenData` -> `writeScreenOpeningData`, now takes the opening player as well
- Improved documentation and added usage examples.

Juuxel committed:
Replace the content of links in FabricHandledScreens for readability

Juuxel committed:
Merge remote-tracking branch 'upstream/1.16' into patch-8

Juuxel committed:
Sync with FabLabs

Juuxel committed:
Add test mod

Juuxel commented:
Alright, I've updated the PR with some new names (detailed in the description) and a test mod.

Juuxel committed:
I suppose IInv shouldn't have CC0 anymore :eyes:

Player3324 reviewed (approved):
So much better with the new names! Thanks!

Juuxel committed:
Improve logging

- Networking now uses a logger instead of an exception
- The loggers are now suffixed with their side
- Fixed the client unregistered screen message using the title instead of the ID

Juuxel commented:
@sfPlayer1 Just asking your opinion on one last name change: do you think the handler registration methods could be just `register`? Those would be cleaner IMO, since the functional interface can be inferred from parameters.

Juuxel committed:
Merge branch '1.16' into patch-8

Juuxel committed:
Fix test mod on 20w21a

Juuxel committed:
Fix ExtendedScreenHandlerFactory javadoc referencing old names

Player3324 commented:
I think the current form is a bit better, it is a bit more obvious what is happening and IIRC IDE hinting+completion for these overloads would be rather poor when using lambdas.

Juuxel commented:
Yeah, I've seen especially the Kotlin code completion get buggy when there are multiple Java functional interfaces as overloads.

modmuss50 commented:
I am fairly busy this weekend, but I did plan to review this by taking it for a test ride in TechReborn. I still plan on doing this. Please keep chasing me if I forget.

i509VCB commented:
No rush modmuss :)

modmuss50 commented:
Just ported TR to use this, it does seem a bit more akward to use than the previous API due to it trying to fit into what vanilla does.


Before I could open the screen handler like so:

```java
ContainerProviderRegistry.INSTANCE.openContainer(identifier, player, packetByteBuf -> packetByteBuf.writeBlockPos(pos));
```

As far as I can see this is the replacement:

```java
			player.openHandledScreen(new ExtendedScreenHandlerFactory() {
				@Override
				public void writeScreenOpeningData(ServerPlayerEntity serverPlayerEntity, PacketByteBuf packetByteBuf) {
					packetByteBuf.writeBlockPos(pos);
				}

				@Override
				public Text getDisplayName() {
					return new LiteralText("What is this for?");
				}

				@Nullable
				@Override
				public ScreenHandler createMenu(int syncId, PlayerInventory inv, PlayerEntity player) {
					PacketByteBuf buf = new PacketByteBuf(Unpooled.buffer());
					buf.writeBlockPos(pos);
					return screenHandlerFactory.create(syncId, inv, buf);
				}
			});
```

with screenHandlerFactory being : 
```java
	private ScreenHandlerRegistry.ExtendedClientHandlerFactory<BuiltScreenHandler> getScreenHandlerFactory() {
		return (syncId, playerInventory, packetByteBuf) -> {
			final BlockEntity blockEntity = playerInventory.player.world.getBlockEntity(packetByteBuf.readBlockPos());
			BuiltScreenHandler screenHandler = ((BuiltScreenHandlerProvider) blockEntity).createScreenHandler(syncId, playerInventory.player);

			//Set the screen handler type, not ideal but works lol
			screenHandler.setType(screenHandlerType);

			return screenHandler;
		};
	}
```

ScreenRegistry.Factory doesnt give me the packetByteBuf like ContainerFactory used to, making it slighly annoying but not a major issue.

I have pushed the changes here:

https://github.com/TechReborn/TechReborn/commit/3e2753affb597d9a9487a273ff4e97a04e35898e

and here: 

https://github.com/TechReborn/RebornCore/commit/c689b15e60a935ec56d7316471f9ad23f32ba71f

This API does seem to be able to cover all of the old uses but like I said it does feel slighly limited by vanilla, it wasnt too hard for me to move over to use it as I already had everything abstracted away, however I can see it being a pain for mods that havent done this.

modmuss50 commented on fabric-screen-handler-api-v1/src/main/java/net/fabricmc/fabric/api/client/screenhandler/v1/package-info.java:
```diff
@@ -0,0 +1,20 @@
+/**
+ * API for working with screen handlers on the client.
+ */
+package net.fabricmc.fabric.api.client.screenhandler.v1;
```
Are these needed? I dont think we have them anywhere else in api?

Juuxel replied:
They make the javadoc nicer to browse since you can see a description of the package in addition to the name.

modmuss50 replied:
Ok, thats good then.

Juuxel commented:
For clean opening code, you should just implement `ExtendedScreenHandlerFactory` on your block entity and use `state.createScreenHandlerFactory` (see `BlockWithEntity` subclasses like chests). Creating anonymous classes seems like a hack to pretend the module is like containers-v0. Especially code like `screenHandlerFactory.create(syncId, inv, buf);` looks like a hacky workaround.

> `screenHandler.setType(screenHandlerType);`

Why are you setting the type manually? :concern: You should just pass it to the `ScreenHandler` constructor.

> ScreenRegistry.Factory doesnt give me the packetByteBuf like ContainerFactory used to, making it slighly annoying but not a major issue.

Screens not having an exposed packet buffer is intended. You're supposed to read all data into your handler class and use that in your screen. This wouldn't be too hard to add, but I don't really want to add yet another simple/extended split.

> ```java
> @Override
> public Text getDisplayName() {
>     return new LiteralText("What is this for?");
> }
> ```

That's the screen title that is passed to `Screen.<init>`.

modmuss50 commented:
This should be good for last call if its ready.

Juuxel commented:
It should be ready for last call, I don't think there's anything to change anymore.

Edit: I think I'll test it on the latest snapshot, though, just to be sure.

Juuxel committed:
Merge remote-tracking branch 'upstream/1.16' into patch-8

Juuxel committed:
Update test mod to 1.16-pre2

Just a simple mapping update. :)

Juuxel committed:
Fix incorrect javadoc references to screen-handler-api in containers-v0

Juuxel committed:
Remove Minecraft exclusion from build.gradle (thanks to Loom 0.4) + testmod change

Resource loader v0 is now testmodRuntimeOnly as it's not needed for compiling.

i509VCB commented:
It has been a week in last call with this PR. Can you update to 1.16-pre5 before merge?

Juuxel commented:
I don't think there's anything to update, but I can't really check since there's a mixin problem with the test mod (see #api on discord).
